dcos_statsd input plugin

philipnrmn opened the pull request:
This PR adds the dcos_statsd input plugin.

The plugin gets statsd metrics from tasks running under mesos on every call to Gather(). It decorates them with their container ID and passes them into the pipeline. 

Resolves [DCOS_OSS-3718](https://jira.mesosphere.com/browse/DCOS_OSS-3718)

philipnrmn committed:
Add readme for dcos_statsd

philipnrmn committed:
Add skeleton for dcos_statsd plugin

philipnrmn committed:
Add stub command API

philipnrmn committed:
Outline structure of dcos_statsd

philipnrmn committed:
Expand response codes for post /container

philipnrmn committed:
Move container.go to dedicated package

philipnrmn committed:
Introduce containers.Controller type, implemented in DCOSStatsd

Update comments for controller funcs

philipnrmn committed:
Pass statsd instance by reference (contains lock)

philipnrmn committed:
Add tests for Start and Stop

philipnrmn committed:
Add test for empty containers

philipnrmn committed:
Fill out API

philipnrmn committed:
Simplify table-driven test

Remove stray API_TEST var

philipnrmn committed:
Add simplified containers dir

philipnrmn committed:
Load cached containers from disk on start

philipnrmn committed:
Refactor loadContainers to use AddContainer

philipnrmn committed:
Stop statsd servers on dcos_statsd stop

philipnrmn committed:
Correct parameter name in api

philipnrmn committed:
Add tests for API

philipnrmn committed:
Add Gather method and tests

philipnrmn committed:
Add custom accumulator for container_id tags

Remove stray method

philipnrmn committed:
Remove containers

philipnrmn committed:
Write containers to disk

philipnrmn committed:
Add dcos_statsd to all.go

philipnrmn committed:
Remove persisted containers from disk

philipnrmn committed:
Remove extraneous sleep

philipnrmn committed:
Listen on a unix socket optionally

philipnrmn committed:
remove accidentally commited swp file

philipnrmn committed:
Add spartan IP as default statsd host

philipnrmn committed:
Bring README into line with default config

branden commented on plugins/inputs/dcos_statsd/dcos_statsd.go:
```diff
@@ -0,0 +1,331 @@
+			if err := c.Server.Gather(cacc); err != nil {
+				log.Printf("E! Error gathering statsd from %s: %s", c.Id, err)
+			}
+		}(ctr)
```
`ctr` is passed as an arg to this function, but it closes over `acc` and `wg`, and I can't tell if there's a reason for the difference. Since we're calling this function as soon as we define it, could it also close over `ctr`?

philipnrmn replied:
there's only one instance of `acc` and one instance of `wg` here, but `ctr` is different for each goroutine. Unfortunately it's also a reference, so closing over `ctr` could (and does) end up sending some references twice or more and some not at all.

branden commented on plugins/inputs/dcos_statsd/README.md:
````diff
@@ -0,0 +1,66 @@
+```
+$ telegraf --config dcos.conf --input-filter dcos_statsd --test
+* Plugin: dcos_statsd
+  cpus,host=172.17.8.102,container_id=12377985-615c-4a1a-a491-721ce7cd807a database.rows.written=12345,database.iops=100
````
This example is missing the `metric_type` tag.

branden commented on plugins/inputs/dcos_statsd/README.md:
```diff
@@ -0,0 +1,66 @@
+
+All metrics have the following tags:
+
+ - container_id
```
This is missing `metric_type`.

philipnrmn committed:
Avoid data race waiting for server to start

philipnrmn committed:
Wait for stats to be present before testing tags

philipnrmn committed:
Fixup govet issue

philipnrmn committed:
Bump timeout to 5s

philipnrmn committed:
Move sleep inside for loop

philipnrmn committed:
Add metric_type tag to README

philipnrmn committed:
Bump test timeout to 30s

branden commented:
This looks like what's causing the build to fail: https://github.com/golang/lint/issues/397 Not sure why only this PR is running into it though.

philipnrmn committed:
Add dcos_statsd to README

philipnrmn commented:
I added a small commit to trigger the tests, maybe it was just a network hiccup...

branden commented:
`dcos_statsd` is causing Telegraf to die shortly after starting on https://github.com/dcos/dcos/pull/3366. Nothing interesting in logs unfortunately, even with debug enabled -- I had to figure out the culprit by commenting out plugins in config.

```
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal systemd[1]: Started Telegraf: collects and reports metrics.
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + set -o nounset
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + cluster_id_file=/var/lib/dcos/cluster-id
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + '[' '!' -f /var/lib/dcos/cluster-id ']'
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: ++ cat /var/lib/dcos/cluster-id
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + cluster_id=2227154d-d5cf-4ac1-aec7-09dd11c0668e
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: ++ /opt/mesosphere/bin/detect_ip
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + node_private_ip=10.0.6.219
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + export DCOS_CLUSTER_ID=2227154d-d5cf-4ac1-aec7-09dd11c0668e
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + DCOS_CLUSTER_ID=2227154d-d5cf-4ac1-aec7-09dd11c0668e
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + export DCOS_NODE_PRIVATE_IP=10.0.6.219
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + DCOS_NODE_PRIVATE_IP=10.0.6.219
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + export DCOS_MESOS_ID=fake_mesos_id
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + DCOS_MESOS_ID=fake_mesos_id
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: + exec /opt/mesosphere/bin/telegraf --config /opt/mesosphere/etc/telegraf/telegraf.conf --config-directory /opt/mesosphere/etc/telegraf/telegraf.d/
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: 2018-08-30T08:35:05Z D! Attempting connection to output: prometheus_client
Aug 30 08:35:05 ip-10-0-6-219.us-west-2.compute.internal start_telegraf.sh[16426]: 2018-08-30T08:35:05Z D! Successfully connected to output: prometheus_client
Aug 30 08:35:10 ip-10-0-6-219.us-west-2.compute.internal systemd[1]: dcos-telegraf.service: Service hold-off time over, scheduling restart.
Aug 30 08:35:10 ip-10-0-6-219.us-west-2.compute.internal systemd[1]: Stopped Telegraf: collects and reports metrics.
Aug 30 08:35:10 ip-10-0-6-219.us-west-2.compute.internal systemd[1]: Starting Telegraf: collects and reports metrics...
```

philipnrmn commented:
It might well be the lack of existence of the containers file... I'll go look.

philipnrmn committed:
Modify assertions to test that one packet arrived

We were previously insisting that all ten packets arrive, be counted and
tagged. We now have a lower standard that still tests that everything is
functioning correctly.

philipnrmn committed:
Split out UDP-specific tests to separate tag

philipnrmn commented:
I moved the UDP-dependent tests into a separate file, as discussed.

philipnrmn commented:
@branden and that's fixed it. I think our CI environment is not UDP packet friendly.

branden reviewed (approved):
:ship:
